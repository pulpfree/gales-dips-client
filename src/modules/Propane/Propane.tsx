import { ContentContainer } from '../Components/ContentContainer'

export function Propane() {
  return (
    <ContentContainer>
      <h2>Propane</h2>
    </ContentContainer>
  )
}