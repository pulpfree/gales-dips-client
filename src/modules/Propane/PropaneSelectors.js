import React, { Component } from 'react'
import PropTypes from 'prop-types'

import classNames from 'classnames'
import moment from 'moment'
import { withRouter } from 'react-router'

import ArrowBack from '@material-ui/icons/ArrowBack'
import ArrowForward from '@material-ui/icons/ArrowForward'
import Button from '@material-ui/core/Button'
import FormControl from '@material-ui/core/FormControl'
import { withStyles } from '@material-ui/core/styles'

import MomentUtils from 'material-ui-pickers/utils/moment-utils'
import MuiPickersUtilsProvider from 'material-ui-pickers/utils/MuiPickersUtilsProvider'
import DatePicker from 'material-ui-pickers/DatePicker'

// import { extractPathParts } from '../../utils/utils'
import { STD_DATE_FORMAT as dateFormat } from '../../config/constants'

class PropaneSelectors extends Component {
  state = {
    nextDisabled: true,
    selectedDate: moment(),
  }

  componentDidMount = () => {
    // const pathPrts = extractPathParts(this.props.location.pathname, 2)
    // if (!pathPrts) return

    const { match: { params } } = this.props
    if (!params.date) return
    const date = params.date
    // console.log('match: ', this.props.match.params)
    // console.log('date: ', date)
    // this.setState({ selectedDate: date }, this.handleGetEntry)
    this.setState({ selectedDate: moment(date) }, this.setNextDisabled)

    /* const { history, match } = this.props
    const { selectedDate } = this.state
    const dte = selectedDate.format(dateFormat)
    const uri = `${match.url}/${dte}`
    history.push(uri) */
  }

  handleDateChange = (date) => {
    this.setState({ selectedDate: date }, this.handleGetEntry)
  }

  handleNextPrevDate = (val) => {
    const dte = this.state.selectedDate
    if (val === 'p') {
      dte.subtract(1, 'days')
    } else if (val === 'n') {
      dte.add(1, 'days')
    }
    this.setState({ selectedDate: dte }, this.handleGetEntry)
  }

  handleGetEntry = () => {
    this.setNextDisabled()
    const { history } = this.props
    const { selectedDate } = this.state

    const dte = selectedDate.format(dateFormat)
    // const uri = `${match.url}/${dte}`
    const uri = `/propane/${dte}`
    history.push(uri)
  }

  setNextDisabled = () => {
    const { selectedDate } = this.state
    const today = moment().format(dateFormat)
    this.setState({ nextDisabled: !selectedDate.isBefore(today) })
  }

  render() {
    const { classes } = this.props
    const { nextDisabled, selectedDate } = this.state

    return (
      <div className={classes.container}>
        <div className={classes.selectRow}>

          <div className={classNames([classes.cell], [classes.calendar])}>
            <FormControl className={classes.formControl}>
              <MuiPickersUtilsProvider utils={MomentUtils}>
                <DatePicker
                  autoOk
                  disableFuture
                  format="MMM D, YYYY"
                  label="Date"
                  onChange={this.handleDateChange}
                  value={selectedDate}
                />
              </MuiPickersUtilsProvider>
            </FormControl>
          </div>

          <div className={classes.cell}>
            <Button
              className={classes.navButton}
              color="secondary"
              onClick={() => this.handleNextPrevDate('p')}
              variant="fab"
            >
              <ArrowBack />
            </Button>
            <Button
              className={classes.navButton}
              color="secondary"
              disabled={nextDisabled}
              onClick={() => this.handleNextPrevDate('n')}
              variant="fab"
            >
              <ArrowForward />
            </Button>
          </div>

        </div>
      </div>
    )
  }
}

PropaneSelectors.propTypes = {
  classes: PropTypes.object.isRequired,
  history: PropTypes.object.isRequired,
  location: PropTypes.object.isRequired,
  match: PropTypes.object.isRequired,
}

const styles = theme => ({
  cell: {
    flex: 'flex-grow',
    alignSelf: 'flex-end',
  },
  container: {
    marginBottom: theme.spacing.unit,
  },
  formControl: {
    margin: theme.spacing.unit,
    minWidth: 160,
  },
  navButton: {
    margin: theme.spacing.unit,
    height: theme.spacing.unit * 4.5,
    width: theme.spacing.unit * 4.5,
    marginRight: 0,
  },
  selectRow: {
    display: 'inline-flex',
    flexDirection: 'row',
    marginBottom: theme.spacing.unit,
  },
})

const comp = withRouter(PropaneSelectors)
export default withStyles(styles)(comp)
