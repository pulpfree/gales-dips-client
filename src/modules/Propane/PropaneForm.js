import React, { Component } from 'react'
import PropTypes from 'prop-types'

import gql from 'graphql-tag'
import { graphql, Mutation } from 'react-apollo'

import classNames from 'classnames'

import Button from '@material-ui/core/Button'
import Delete from '@material-ui/icons/Delete'
import Input from '@material-ui/core/Input'
import Save from '@material-ui/icons/Save'
import Typography from '@material-ui/core/Typography'
import { withStyles } from '@material-ui/core/styles'

import PropaneSelectors from './PropaneSelectors'
import { extractPathParts, dateToInt } from '../../utils/utils'


const DELIVERY_QUERY = gql`
query PropaneDelivery($date: Int!) {
  propaneDelivery(date: $date) {
    date
    litres
  }
}
`

const CREATE_DELIVERY = gql`
mutation CreatePropaneDelivery($fields: PropaneDeliverInput!) {
  createPropaneDelivery(input: $fields) {
    ok
    nModified
  }
}
`

const REMOVE_DELIVERY = gql`
mutation RemovePropaneDelivery($fields: PropaneRemoveDeliverInput!) {
  removePropaneDelivery(input: $fields) {
    ok
    nModified
  }
}
`

class PropaneForm extends Component {

  state = {
    litres: '',
  }

  componentDidUpdate = (prevProps) => {
    const { data, location } = this.props
    // When changing dates and no delivery, remove litres
    if (!data.propaneDelivery && location !== prevProps.location) {
      this.setState({litres: ''})
    }
    // First time we get delivery set litres
    if (prevProps.data && !prevProps.data.propaneDelivery && data.propaneDelivery) {
      this.setState({litres: data.propaneDelivery.litres})
    }
    // When moving between dates and we have deliver set litres
    if (prevProps.data && data.propaneDelivery && prevProps.data.propaneDelivery !== data.propaneDelivery) {
      this.setState({litres: data.propaneDelivery.litres})
    }
    /*if (location !== prevProps.location) {
      console.log('data.propaneDelivery: ', data.propaneDelivery)
    }*/
  }

  createDeliveryParams = () => {
    const prts = extractPathParts(this.props.location.pathname)
    if (!prts) return null
    const date = dateToInt(prts[0])
    if (!this.state.litres) return null
    return {
      fields: {
        date,
        litres: Number(this.state.litres),
      },
    }
  }

  handleChange = prop => event => {
    this.setState({ [prop]: parseInt(event.target.value, 10) })
  }

  removeDeliveryParams = () => {
    const prts = extractPathParts(this.props.location.pathname)
    if (!prts) return null
    const date = dateToInt(prts[0])
    if (!this.state.litres) return null
    return {
      fields: {
        date,
      },
    }
  }

  removeLitresState = () => {
    this.setState({litres: ''})
  }

  render() {

    const { classes } = this.props
    const { litres } = this.state

    return (
      <div className={classes.container}>
        <Typography
            gutterBottom
            variant="title"
        >Delivery</Typography>
        <PropaneSelectors />

        <div className={classes.dataRow}>
          <div className={classes.dataCell}>
            <Input
                autoFocus
                name="litres"
                onChange={this.handleChange('litres')}
                placeholder="Litres"
                type="number"
                value={litres}
            />
          </div>
          <Mutation
              mutation={REMOVE_DELIVERY}
              onCompleted={this.removeLitresState}
              variables={this.removeDeliveryParams()}
          >
            {(removeDelivery, { loading, error }) => (
              <div className={classNames([classes.dataCell], [classes.narrowCell])}>
                <Button
                    className={classes.delButton}
                    color="secondary"
                    disabled={!this.state.litres}
                    onClick={removeDelivery}
                >
                  <Delete style={{padding: 0}} />
                </Button>
                {loading && <div>Processing...</div>}
                {error && <div>Error :( Please try again</div>}
              </div>
            )}
          </Mutation>
        </div>

        <Mutation
            mutation={CREATE_DELIVERY}
            variables={this.createDeliveryParams()}
        >
          {(createDelivery, { loading, error }) => (
            <div>
              <Button
                  className={classes.submitButton}
                  color="primary"
                  disabled={!this.state.litres}
                  onClick={createDelivery}
                  variant="raised"
              >
                <Save className={classNames(classes.leftIcon, classes.iconSmall)} />
                Save Propane Delivery
              </Button>
              {loading && <div>Processing...</div>}
              {error && <div>Error :( Please try again</div>}
            </div>
          )}
        </Mutation>
      </div>
    )
  }
}

PropaneForm.propTypes = {
  classes:  PropTypes.object.isRequired,
  data:     PropTypes.object,
  location: PropTypes.object.isRequired,
}

const styles =  theme => ({
  alignRight: {
    textAlign: 'right',
  },
  alignCenter: {
    textAlign: 'center',
  },
  container: {
    display:        'flex',
    flexDirection:  'column',
    fontFamily:     theme.typography.fontFamily,
    width:          theme.spacing.unit * 40,
    margin: 'auto',
  },
  dataCell: {
    alignSelf:      'flex-end',
    flex:           1,
    padding:        theme.spacing.unit,
    paddingBottom:  theme.spacing.unit * 2,
  },
  dataRow: {
    display:        'inline-flex',
    flexDirection:  'row',
  },
  delButton: {
    minWidth: 0,
  },
  leftIcon: {
    marginRight: theme.spacing.unit,
  },
  iconSmall: {
    fontSize: 20,
  },
  submitButton: {
    margin: theme.spacing.unit,
    marginTop: theme.spacing.unit * 3,
  },
})

const styledForm = withStyles(styles)(PropaneForm)

export default graphql(DELIVERY_QUERY, {
  skip: props => props.location.pathname.split('/').length <= 2,
  options: (props) => {
    const prts = extractPathParts(props.location.pathname)
    return ({
      variables: {
        date: prts ? dateToInt(prts[0]) : null,
      },
    })
  },
})(styledForm)
