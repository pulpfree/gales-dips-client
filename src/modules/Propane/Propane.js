import React, { Component } from 'react'
import PropTypes from 'prop-types'

import Divider from '@material-ui/core/Divider'
import Paper from '@material-ui/core/Paper'
import Typography from '@material-ui/core/Typography'
import { withStyles } from '@material-ui/core/styles'

import Header from '../Header/Header'
import PropaneSelectors from './PropaneSelectors'
import { styles as ms } from '../../styles/main'

import PropaneForm from './PropaneForm'

class Propane extends Component {

  render() {

    const { classes, history, location, match } = this.props

    return (
      <div>
        <Header history={history} />
        <Paper className={classes.paper}>
          <Typography
              gutterBottom
              variant="headline"
          >Propane Entries</Typography>
          <Divider /><br />
          <PropaneSelectors
              history={history}
              location={location}
              match={match}
          />
          <div style={{marginTop: 20}}>
            <PropaneForm location={location} />
          </div>
        </Paper>
      </div>
    )
  }
}

Propane.propTypes = {
  classes:  PropTypes.object.isRequired,
  history:  PropTypes.object.isRequired,
  location: PropTypes.object.isRequired,
  match:    PropTypes.object.isRequired,
}

export default withStyles(ms)(Propane)