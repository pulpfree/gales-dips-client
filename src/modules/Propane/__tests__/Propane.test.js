/* eslint no-undef: "off" */

import React from 'react'
import { shallow } from 'enzyme'

import Propane from '../Propane'

describe('Propane', () => {
  let mountedC
  let props
  const reportC = () => {
    if (!mountedC) {
      mountedC = shallow(<Propane {...props} />)
    }
    return mountedC
  }

  beforeEach(() => {
    props = {
      history: {},
      location: {},
      match: {},
    }
    mountedC = undefined
  })

  it('matches snapshot', () => {
    expect(reportC()).toMatchSnapshot()
  })
})
