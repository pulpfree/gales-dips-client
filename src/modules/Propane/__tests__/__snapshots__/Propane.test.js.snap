// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Propane matches snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <WithStyles(Propane)
    history={Object {}}
    location={Object {}}
    match={Object {}}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "classes": Object {
        "flex": "Propane-flex-2",
        "mainContainer": "Propane-mainContainer-3",
        "paper": "Propane-paper-5",
        "popperClose": "Propane-popperClose-6",
        "root": "Propane-root-1",
        "titleLink": "Propane-titleLink-4",
      },
      "history": Object {},
      "location": Object {},
      "match": Object {},
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "classes": Object {
          "flex": "Propane-flex-2",
          "mainContainer": "Propane-mainContainer-3",
          "paper": "Propane-paper-5",
          "popperClose": "Propane-popperClose-6",
          "root": "Propane-root-1",
          "titleLink": "Propane-titleLink-4",
        },
        "history": Object {},
        "location": Object {},
        "match": Object {},
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
